tinync/tinync.c

```diff
@@ -14,40 +14,52 @@ enum {
 struct cr {
     void *label;
     int status;
-    void *local; /* private local storage */
+    void *arg;
 };
 
-#define cr_init()                           \
-    {                                       \
-        .label = NULL, .status = CR_BLOCKED \
+#define cr_context_name(name) __cr_context_##name
+#define cr_context(name) struct cr cr_context_name(name)
+#define cr_context_init(argptr)                            \
+    {                                                      \
+        .label = NULL, .status = CR_BLOCKED, .arg = argptr \
     }
-#define cr_begin(o)                     \
-    do {                                \
-        if ((o)->status == CR_FINISHED) \
-            return;                     \
-        if ((o)->label)                 \
-            goto *(o)->label;           \
+
+#define cr_func_name(name) __cr_func_##name
+#define cr_define(name) static void cr_func_name(name)(struct cr * ctx)
```

Comment:
Let's respect the unified C-style function declaration as possible.

Reply:
That is, avoid the indirection such as `cr_define`.

Reply:
I intent to keep `cr_define` since its main purpose is to avoid misusing to coroutine controlling macros, but with some proper modifications, using to those macros could become more C-style from now. Following is my plan to extend `cr_define`:
* modify the content of `cr_define` as `#define cr_define(name, ...) cr_func_name(name)(struct cr * ctx, ## __VA_ARGS_)`, which makes programmers define a coroutine function as `type cr_define(coroutine_name, /* argumnet list */)` that allows programmers to specify custom argument list for specific function.
* Respond to previous change, content of `cr_run` should be modified as `#define cr_run(name, ...) cr_func_name(name)(&cr_context_name(name), ##__VA_ARGS__)`.
These modifications makes those indirect macros become more flexible while keeping programmers from misusing controlling macros.

Reply:
We can use certain prototype macros instead. Let's say, `cr_proto`, which generates the appropriate function prototype for coroutines.
See https://stackoverflow.com/questions/57779123/whats-the-point-of-a-prototype-macro-that-merely-expands-to-its-arguments

Reply:
You mean to use macro to generate argument list instead of entire declaration of function?

Reply:
Assume `cr_proto` is defined. I expect the use as following:
```c
static void cr_proto(fetch_data, int x)
{
    ...
}
```

Then, the actual declaration would be:
```c
static void __cr_func_fetch_data(int x)
{
    ...
}
```

Reply:
Thank for your explanation, I think I figure out what you expect for. I will modify those macros to fulfill that.

Reply:
By the way, coroutine is indeed a common technique. You can check another project which applies the coroutine technique: https://github.com/sysprog21/timeout/blob/master/timeout.c#L576

Reply:
> By the way, coroutine is indeed a common technique. You can check another project which applies the coroutine technique: https://github.com/sysprog21/timeout/blob/master/timeout.c#L576

I notice that a `switch case` statement is used to control the execution point, rather than labels as values in this project. I'll spend time to figure out pros and cons between these techniques.

Reply:
Comparisons:
1. Local continuation based on switch/case and line numbers.
  * Pros: works with any C99 compiler, most simple implementation.
  * Cons: does not allow more than one label per line, does not preserve local variables.
2. Local continuation based on goto label references.
  * Pros: works with all control structures. slightly faster.
   * Cons: requires GCC or Clang, does not preserve local variables.

Feel free to contribute to [sysprog21/timeout](https://github.com/sysprog21/timeout).